### DISCUSSION TITLE
API for /users/{user}/progress/watched/activity?hide_completed=true

### DISCUSSION THREAD

**Discussion by mpfc75** [trakt/trakt-api#300] [Q&A] [answered]
On the Trakt.tv website I came across this address:
https://trakt.tv/users/{user}/progress/watched/activity?hide_completed=true

It gives a list of the next episodes to watch for the shows the user watches.

On Postman I tried the command https://api.trakt.tv/users/{user}/progress/watched/activity?hide_completed=true
and got back last episode and next episode info for all shows that have a next episode. 

I don't see this api documented on https://trakt.docs.apiary.io/ am I just missing it?  It would be very useful if it could be used.

Here is part of the JSON I get back:
```
[
    {
        "show": {
            "title": "The Crown",
            "year": 2016,
            "ids": {
                "trakt": 104578,
                "slug": "the-crown",
                "tvdb": 305574,
                "imdb": "tt4786824",
                "tmdb": 65494,
                "tvrage": 42736
            }
        },
        "progress": {
            "aired": 40,
            "completed": 16,
            "last_watched_at": "2021-07-06T04:55:45.000Z",
            "reset_at": null,
            "next_episode": {
                "season": 2,
                "number": 7,
                "title": "Matrimonium",
                "ids": {
                    "trakt": 2705842,
                    "tvdb": 6276532,
                    "imdb": "tt6955528",
                    "tmdb": 1401395,
                    "tvrage": null
                }
            },
            "last_episode": {
                "season": 2,
                "number": 6,
                "title": "Vergangenheit",
                "ids": {
                    "trakt": 2705841,
                    "tvdb": 6276531,
                    "imdb": "tt6837072",
                    "tmdb": 1401394,
                    "tvrage": null
                }
            }
        }
    }, ... etc.
```

**Comment by rectifyer** [comment, marked as answer]
For performance reasons, this API method shouldn't be publicly available. It looks like a recent update exposed it, when it shouldn't have been. The API has been corrected and the API docs are correct since this method isn't supported.

**Comment by mpfc75** [comment]
Thanks for the quick reply!

Can you help me understand the best alternative for getting the next episode to watch for all of a users shows?  Unless I am missing something, the alternative is to have a list of all shows with unwatched episodes (cached or from one or more api calls).  Then, for each show, preform /shows/{show}/progress/watched.

The /shows/{show}/progress/watched returns not only the next and last episode info but additionally data for each episode in the show (watched or not).  The older the show, the more data gets transferred.

It seems that this would result in many more api calls (1 per show plus perhaps calls to get list of show) and the data returned would be much larger than the succinct /users/{user}/progress/watched/activity call.

Is there a better way to use the api's to obtain the same information?

I appreciate your help!

**Reply by rectifyer** [reply]
I recommend doing this comparison locally instead of hitting the API for every show. Presumably you already are caching info about the shows and seasons. You can then use https://trakt.docs.apiary.io/#reference/sync/get-watched/get-watched and do local comparisons to calculate the progress for each show you need.

**Reply by mpfc75** [reply]
Yes, I was doing that earlier, but I was trying to account for updates that might have been made on another device, app, or directly on the website. I suppose I'll change my philosophy and just worry about the device/app and have a way to refresh from time to time.  When I came across the undocumented (and inadvertently exposed) api I thought I had found a workable solution.

**Reply by rectifyer** [reply]
I'd also recommend using https://trakt.docs.apiary.io/#reference/sync/last-activities/get-last-activity to know if the user has updated data. That's a quick way to know if any new episodes were watched.
